imu_integrator/plugin.cpp

```diff
@@ -0,0 +1,424 @@
+			real_time,
+		});
+    }
+
+	// Select IMU readings based on timestamp similar to how OpenVINS selects IMU values to propagate
+	std::vector<imu_type> select_imu_readings(const std::vector<imu_type>& imu_data, double time_begin, double time_end) {
+		std::vector<imu_type> prop_data;
+		if (imu_data.size() < 2) {
+			return prop_data;
+		}
+
+		for (unsigned i = 0; i < imu_data.size()-1; i++) {
+
+			// If time_begin comes inbetween two IMUs (A and B), interpolate A forward to time_begin
+			if (imu_data.at(i+1).timestamp > time_begin && imu_data.at(i).timestamp < time_begin) {
+				imu_type data = interpolate_imu(imu_data.at(i), imu_data.at(i+1), time_begin);
+				prop_data.push_back(data);
+				continue;
+			}
+
+			// IMU is within time_begin and time_end
+			if (imu_data.at(i).timestamp >= time_begin && imu_data.at(i+1).timestamp <= time_end) {
+				prop_data.push_back(imu_data.at(i));
+				continue;
+			}
+
+			// IMU is past time_end
+			if (imu_data.at(i+1).timestamp > time_end) {
+				imu_type data = interpolate_imu(imu_data.at(i), imu_data.at(i+1), time_end);
+				prop_data.push_back(data);
+				break;
+			}
+		}
+
+		// Loop through and ensure we do not have an zero dt values
+		// This would cause the noise covariance to be Infinity
+		for (size_t i = 0; i < prop_data.size()-1; i++) {
+			if (std::abs(prop_data.at(i+1).timestamp-prop_data.at(i).timestamp) < 1e-12) {
+				prop_data.erase(prop_data.begin()+i);
+				i--;
+			}
+		}
+
+		return prop_data;
+	}
+
+	// For when an integration time ever falls inbetween two imu measurements (modeled after OpenVINS)
+	static imu_type interpolate_imu(const imu_type imu_1, imu_type imu_2, double timestamp) {
+		imu_type data;
+		data.timestamp = timestamp;
+
+		double lambda = (timestamp - imu_1.timestamp) / (imu_2.timestamp - imu_1.timestamp);
+		data.am = (1 - lambda) * imu_1.am + lambda * imu_2.am;
+		data.wm = (1 - lambda) * imu_1.wm + lambda * imu_2.wm;
+
+		return data;
+	}
+
+	void predict_mean_rk4(Eigen::Vector4d quat, Eigen::Vector3d pos, Eigen::Vector3d vel, double dt,
+                                  const Eigen::Vector3d &w_hat1, const Eigen::Vector3d &a_hat1,
+                                  const Eigen::Vector3d &w_hat2, const Eigen::Vector3d &a_hat2,
+                                  Eigen::Vector4d &new_q, Eigen::Vector3d &new_v, Eigen::Vector3d &new_p) {
+
+		// Pre-compute things
+		Eigen::Vector3d w_hat = w_hat1;
+		Eigen::Vector3d a_hat = a_hat1;
+		Eigen::Vector3d w_alpha = (w_hat2-w_hat1)/dt;
+		Eigen::Vector3d a_jerk = (a_hat2-a_hat1)/dt;
+
+		// y0 ================
+		Eigen::Vector4d q_0 = quat;
+		Eigen::Vector3d p_0 = pos;
+		Eigen::Vector3d v_0 = vel;
+
+		// k1 ================
+		Eigen::Vector4d dq_0 = {0,0,0,1};
+		Eigen::Vector4d q0_dot = 0.5*Omega(w_hat)*dq_0;
+		Eigen::Vector3d p0_dot = v_0;
+		Eigen::Matrix3d R_Gto0 = quat_2_Rot(quat_multiply(dq_0,q_0));
+		Eigen::Vector3d v0_dot = R_Gto0.transpose()*a_hat-Eigen::Matrix<double,3,1>(0.0, 0.0, 9.81);
```

Comment:
Put gravity in a variable and use it everywhere else. Plus, doesn't gravity come from SLAM?

Reply:
In frustration of finding the flipped gravity bug I just hardcoded the gravity vector in RK4. I think this is a better solution than assuming that the gravity coming from SLAM is negative and then negating it here.

Reply:
Yeah, I agree. We shouldn't assume what kind of gravity vector SLAM will publish. However, sometime in the future we should standardize the interface even more and require all SLAMs to publish gravity in a specific way.